File: implementations/java/AdjacencyList.java
import java.util.*;

public class AdjacencyList {
    private Map<Integer, List<Integer>> adjList;

    public AdjacencyList() {
        this.adjList = new HashMap<>();
    }

    public void addVertex(int vertex) {
        adjList.putIfAbsent(vertex, new ArrayList<>());
    }

    public void addEdge(int source, int destination, boolean bidirectional) {
        addVertex(source);
        addVertex(destination);
        adjList.get(source).add(destination);
        if (bidirectional) {
            adjList.get(destination).add(source);
        }
    }
}

File: implementations/go/adjacency_list.go
package main

type GraphList struct {
	AdjacencyList map[int][]int
}

func NewGraphList() *GraphList {
	return &GraphList{
		AdjacencyList: make(map[int][]int),
	}
}

func (g *GraphList) AddVertex(vertex int) {
	if _, exists := g.AdjacencyList[vertex]; !exists {
		g.AdjacencyList[vertex] = []int{}
	}
}

func (g *GraphList) AddEdge(source, destination int, bidirectional bool) {
	g.AddVertex(source)
	g.AddVertex(destination)
	g.AdjacencyList[source] = append(g.AdjacencyList[source], destination)
	if bidirectional {
		g.AdjacencyList[destination] = append(g.AdjacencyList[destination], source)
	}
}
